### PR TITLE
Netsuite Test Goal 80%

### PR DESCRIPTION
**Pull request by NileshPant1999**

### CODE REVIEW

File: tests/test_fyle/test_connector.py [hunk 1]
```diff
@@ -21,6 +21,14 @@ def test_get_of_expenses(add_fyle_credentials):
     assert len(personal_expenses) == 4
     assert len(ccc_expenses) == 2
 
+    expense_group_settings = ExpenseGroupSettings.objects.get(workspace_id=1)
+    expense_group_settings.import_card_credits = False
+
+    personal_expenses = fyle_connector.get_expenses(
```
Comment (ashwin1111) [inline tests/test_fyle/test_connector.py:27]: this would fail after 2nd phase, better to remove it before merging

File: tests/test_fyle/test_views.py [hunk 2]
```diff
@@ -1,12 +1,111 @@
+from os import access
```
Comment (ashwin1111) [inline tests/test_fyle/test_views.py:1]: why is this needed?

File: tests/test_fyle/test_views.py [hunk 3]
```diff
@@ -1,12 +1,111 @@
...
+from urllib import response
```
Comment (ashwin1111) [inline tests/test_fyle/test_views.py:2]: why is this needed?

File: tests/test_fyle/test_views.py [hunk 4]
```diff
@@ -1,12 +1,111 @@
...
+   
+   response = api_client.get(url, {
+      'state': 'READY'
+   })
+   response = json.loads(response.content)
+   assert response == {'count': 1, 'next': None, 'previous': None, 'results': [{'id': 1, 'fund_source': 'PERSONAL', 'description': {'report_id': 'rpuN3bgphxbK', 'fund_source': 'PERSONAL', 'claim_number': 'C/2021/11/R/5', 'employee_email': 'ashwin.t@fyle.in'}, 'response_logs': {'name': None, 'type': 'expenseReport', 'externalId': '03294720937402397402937', 'internalId': '116142'}, 'created_at': '2021-11-15T10:29:07.618062Z', 'exported_at': '2021-11-15T11:02:55.125205Z', 'updated_at': '2021-11-15T11:02:55.125634Z', 'workspace': 1, 'expenses': [1]}]}
+
+
+
+@pytest.mark.django_db(databases=['default'])
```
Comment (ashwin1111) [inline tests/test_fyle/test_views.py:41]: I don't think we need to pass db, it defaults to default

File: tests/test_fyle/test_views.py [hunk 5]
```diff
@@ -1,12 +1,111 @@
...
+
+   expense_group = ExpenseGroup.objects.filter(workspace_id=2).first()
+   url = reverse('expense-group-expenses', 
+      kwargs={
+            'workspace_id': 2,
+            'expense_group_id': expense_group.id
+         }
+      )
+
+   api_client.credentials(HTTP_AUTHORIZATION='Bearer {}'.format(access_token))
```
Comment (ashwin1111) [inline tests/test_fyle/test_views.py:54]: just a question/research item, does it makes sense to add this at top level?

File: tests/test_fyle/test_views.py [hunk 6]
```diff
@@ -27,6 +126,27 @@ def test_fyle_fields_view(api_client, test_connection):
...
+   url = reverse('expense-attributes', 
+      kwargs={
+            'workspace_id': 2
+         }
+      )
+   
+   api_client.credentials(HTTP_AUTHORIZATION='Bearer {}'.format(access_token))
+
+   response = api_client.get(url, {
+      'attribute_type': 'FYLE_CATEGORY'
```
Comment (ashwin1111) [inline tests/test_fyle/test_views.py:142]: is this a custom field?

File: tests/test_fyle/test_views.py [hunk 7]
```diff
@@ -48,23 +168,49 @@ def test_expense_group_id_view(api_client, test_connection):
...
-
+   
    api_client.credentials(HTTP_AUTHORIZATION='Bearer {}'.format(access_token))
+   
+   response = api_client.post(url)
+   assert response.status_code == 200
 
-   response = api_client.get(url)
+   fyle_credentials = FyleCredential.objects.get(workspace_id=1)
+   fyle_credentials.delete()
```
Comment (ashwin1111) [inline tests/test_fyle/test_views.py:189]: what happens to the remaining tests if we delete this? Won't they be affected?

File: tests/test_mappings/conftest.py [hunk 8]
```diff
@@ -1,7 +1,17 @@
...
 import pytest
-from apps.workspaces.models import Configuration, Workspace
-from apps.mappings.models import GeneralMapping
-from apps.fyle.models import ExpenseGroupSettings
\ No newline at end of file
+
+@pytest.fixture
+def create_mapping_settings(db):
+    mapping_setting, _ = MappingSetting.objects.update_or_create(
+                    source_field='CUSTOM_FIELD',
```
Comment (ashwin1111) [inline tests/test_mappings/conftest.py:10]: indent + new line

File: tests/test_mappings/test_helper.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+from time import sleep
+import pytest
+from django_q.models import Schedule
+
+from apps.mappings.helpers import validate_and_trigger_auto_map_employees
+
+
+def test_validate_and_trigger_auto_map_employees(db):
+    validate_and_trigger_auto_map_employees(workspace_id=2)
```
Comment (ashwin1111) [inline tests/test_mappings/test_helper.py:9]: are we not comparing anything here?

File: tests/test_mappings/test_tasks.py [hunk 10]
```diff
@@ -65,4 +82,124 @@ def test_create_fyle_tax_group_payload(db):
...
+    assert schedule.func == 'apps.mappings.tasks.auto_create_cost_center_mappings'
+
+def test_auto_create_category_mappings(db, mocker, add_fyle_credentials, add_netsuite_credentials):
+
+    mocker.patch(
+            'fylesdk.apis.fyle_v1.categories.Categories.post',
+            return_value=[]
+        )
+
+    old_mappings = CategoryMapping.objects.filter(workspace_id=1).count()
```
Comment (ashwin1111) [inline tests/test_mappings/test_tasks.py:169]: We don't use CategoryMapping in QBO I guess, please check.

File: tests/test_mappings/test_tasks.py [hunk 11]
```diff
@@ -65,4 +82,124 @@ def test_create_fyle_tax_group_payload(db):
...
+
+    response = auto_create_category_mappings(workspace_id=1)
+    assert response == []
+
+    categories = DestinationAttribute.objects.filter(workspace_id=1, attribute_type='EXPENSE_CATEGORY').count()
```
Comment (ashwin1111) [inline tests/test_mappings/test_tasks.py:174]: We don't have any attribute of type `EXPENSE_CATEGORY`.

File: tests/test_netsuite/test_connector.py [hunk 12]
```diff
@@ -108,4 +108,30 @@ def test_sync_employee(add_netsuite_credentials):
...
-    assert employee_count == 12
+    assert new_employee_count == 14
+
+@pytest.mark.django_db()
+def test_sync_accounts(add_netsuite_credentials):
+    netsuite_credentials = NetSuiteCredentials.objects.get(workspace_id=1)
+    netsuite_connection = NetSuiteConnector(netsuite_credentials=netsuite_credentials, workspace_id=1)
+
+    accounts_count = DestinationAttribute.objects.filter(attribute_type='ACCOUNT', workspace_id=1).count()
+    assert accounts_count == 123
```
Comment (ashwin1111) [inline tests/test_netsuite/test_connector.py:119]: How do we have control of this count in the fixture, exactly 123?

File: tests/test_netsuite/test_connector.py [hunk 13]
```diff
@@ -108,4 +108,30 @@ def test_sync_employee(add_netsuite_credentials):
...
+
+    new_account_counts = DestinationAttribute.objects.filter(attribute_type='ACCOUNT', workspace_id=1).count()
+    assert new_account_counts == 164
+
+@pytest.mark.django_db()
+def test_sync_expense_categories(add_netsuite_credentials):
+    netsuite_credentials = NetSuiteCredentials.objects.get(workspace_id=1)
+    netsuite_connection = NetSuiteConnector(netsuite_credentials=netsuite_credentials, workspace_id=1)
+
+    expense_categories_count = DestinationAttribute.objects.filter(attribute_type='EXPENSE_CATEGORY', workspace_id=1).count()
```
Comment (ashwin1111) [inline tests/test_netsuite/test_connector.py:131]: How do we have a `EXPENSE_CATEGORY` attribute in Destination?

File: tests/test_netsuite/test_connector.py [hunk 14]
```diff
@@ -108,4 +108,30 @@ def test_sync_employee(add_netsuite_credentials):
...
+    assert expense_categories_count == 33
+
+    netsuite_connection.sync_expense_categories()
+
+    new_expense_categories_count = DestinationAttribute.objects.filter(attribute_type='EXPENSE_CATEGORY', workspace_id=1).count()
```
Comment (ashwin1111) [inline tests/test_netsuite/test_connector.py:136]: How do we have a `EXPENSE_CATEGORY` attribute in Destination?

File: tests/test_netsuite/test_tasks.py [hunk 15]
```diff
@@ -174,13 +176,15 @@ def test_accounting_period_working(create_task_logs, add_netsuite_credentials, a
     configuration.change_accounting_period = True
     configuration.save()
 
-    create_expense_report(expense_group, task_log.id)
```
Comment (ashwin1111) [inline tests/test_netsuite/test_tasks.py:177]: is expense report already taken care? asking since we are replacing them with bills in this PR

File: tests/test_workspaces/test_tasks.py [hunk 16]
```diff
@@ -0,0 +1,23 @@
+from logging import FATAL
```
Comment (ashwin1111) [inline tests/test_workspaces/test_tasks.py:1]: where is this used?

File: tests/test_workspaces/test_tasks.py [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+from apps.workspaces.models import WorkspaceSchedule
+import pytest
+import json
```
Comment (ashwin1111) [inline tests/test_workspaces/test_tasks.py:4]: where is this used?

File: tests/test_workspaces/test_tasks.py [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+    assert ws_schedule.enabled == True
+   
+    schedule_sync(2, False, 0)
+
+    ws_schedule = WorkspaceSchedule.objects.filter(workspace_id=2).last()
+    assert ws_schedule.enabled == False
+
+@pytest.mark.django_db(databases=['default'])
+def test_run_sync_schedule(test_connection, add_fyle_credentials, add_netsuite_credentials):
+    run_sync_schedule(1)
```
Comment (ashwin1111) [inline tests/test_workspaces/test_tasks.py:23]: are we not checking anything here?

File: tests/test_workspaces/test_views.py [hunk 19]
```diff
@@ -1,14 +1,36 @@
...
+        'org_id': 'orf6t6jWUnpx'
+    })
+
+    api_client.credentials(HTTP_AUTHORIZATION='Bearer {}'.format(test_connection.access_token))
+
+    response = api_client.get(url)
+    assert response.status_code == 200
+
+    response = json.loads(response.content)
+    expected_response = get_response_dict('test_workspaces/data.json')
```
Comment (ashwin1111) [inline tests/test_workspaces/test_views.py:31]: why are we maintaining 2 fixtures? one in json and one in python. Can we maintain one consistently?

Comment (NileshPant1999) [reply tests/test_workspaces/test_views.py:31]: yes @ashwin1111 i am planning to fix this in next iteration.

File: tests/test_netsuite/test_tasks.py (hunk 15 above)
Comment (NileshPant1999) [reply tests/test_netsuite/test_tasks.py:177]: yes, expense report was already there in the tests

File: tests/test_netsuite/test_connector.py (hunk 13 above)
Comment (NileshPant1999) [reply tests/test_netsuite/test_connector.py:131]: we sync `expense_category` while syncing Netsuite dimensions.

File: tests/test_netsuite/test_connector.py (hunk 14 above)
Comment (NileshPant1999) [reply tests/test_netsuite/test_connector.py:136]: we sync `expense_category` while syncing Netsuite dimensions.

File: tests/test_netsuite/test_connector.py (hunk 12 above)
Comment (NileshPant1999) [reply tests/test_netsuite/test_connector.py:119]: we get this from test db @ashwin1111

File: tests/test_mappings/test_tasks.py (hunk 11 above)
Comment (NileshPant1999) [reply tests/test_mappings/test_tasks.py:174]: we have @ashwin1111 check `sync_expense_categories` function.

File: tests/test_mappings/test_tasks.py (hunk 10 above)
Comment (NileshPant1999) [reply tests/test_mappings/test_tasks.py:169]: This is Netsuite @ashwin1111

File: tests/test_fyle/test_views.py (hunk 7 above)
Comment (NileshPant1999) [reply tests/test_fyle/test_views.py:189]: @ashwin1111 after running this test the transaction would be rolled back. this won't be having any effect.

File: tests/test_fyle/test_views.py (hunk 6 above)
Comment (NileshPant1999) [reply tests/test_fyle/test_views.py:142]: yes

File: tests/test_fyle/test_connector.py (hunk 1 above)
Comment (NileshPant1999) [reply tests/test_fyle/test_connector.py:27]: @ashwin1111 already fixed this in phase 2 pr

File: tests/test_fyle/test_views.py (hunk 4 above)
Comment (NileshPant1999) [reply tests/test_fyle/test_views.py:39]: will check this and remove this from every test in next pr

File: tests/test_fyle/test_views.py (hunk 5 above)
Comment (NileshPant1999) [reply tests/test_fyle/test_views.py:52]: will do research about it

File: tests/test_workspaces/test_tasks.py (hunk 18 above)
Comment (NileshPant1999) [reply tests/test_workspaces/test_tasks.py:22]: fixed

File: tests/test_mappings/test_tasks.py (hunk 10 above)
Comment (ashwin1111) [reply tests/test_mappings/test_tasks.py:170]: my bad my bad, QBO effect

File: tests/test_mappings/test_tasks.py (hunk 11 above)
Comment (ashwin1111) [reply tests/test_mappings/test_tasks.py:175]: my bad my bad, QBO effect

File: tests/test_netsuite/test_connector.py (hunk 13 above)
Comment (ashwin1111) [reply tests/test_netsuite/test_connector.py:131]: my bad my bad, QBO effect

File: tests/test_netsuite/test_connector.py (hunk 14 above)
Comment (ashwin1111) [reply tests/test_netsuite/test_connector.py:136]: my bad my bad, QBO effect

File: tests/test_netsuite/test_signals.py [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+        name='SRAVAN_DEMO',
+        segment_type="CUSTOM_LIST",
+        script_id="custcol780",
+        internal_id="491",
+        workspace_id=2
+    )
+
+    custom_list = DestinationAttribute.objects.filter(attribute_type='SRAVAN_DEMO').count()
+    assert custom_list == 2
+
```
Comment (Sravanksk) [inline tests/test_netsuite/test_signals.py:37]: remove extra indent (tab-space) here

File: tests/test_netsuite/test_tasks.py [hunk 21]
```diff
@@ -160,12 +160,14 @@ def test_accounting_period_working(create_task_logs, add_netsuite_credentials, a
     for expense in expenses:
         expense.expense_group_id = expense_group.id
         expense.save()
-
+
```
Comment (Sravanksk) [inline tests/test_netsuite/test_tasks.py:163]: rm extra tab space

File: tests/test_workspaces/test_views.py [hunk 22]
```diff
@@ -118,4 +189,76 @@ def test_get_workspace_schedule(api_client, test_connection):
...
+    assert response['message'] == 'NetSuite Credentials not found in this workspace'
+
+    url = reverse('delete-netsuite-credentials', kwargs={
+        'workspace_id': 1
+    })
+
+    response = api_client.delete(url)
+    response = json.loads(response.content)
+
+    assert response['message'] == 'NetSuite credentials deleted'
```
Comment (Sravanksk) [inline tests/test_workspaces/test_views.py:264]: new line

**Review by Sravanksk — APPROVED** [review]
